File: intrinsic/config/environments_test.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/config/environments.h"

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "intrinsic/util/testing/gtest_wrapper.h"

namespace environments {
using ::testing::HasSubstr;

TEST(EnvironmentsTest, FromDomainProd) {
  EXPECT_EQ(FromDomain(kPortalDomainProd).value(), kProd);
  EXPECT_EQ(FromDomain(kAccountsDomainProd).value(), kProd);
  EXPECT_EQ(FromDomain(kAssetsDomainProd).value(), kProd);
}

TEST(EnvironmentsTest, FromDomainStaging) {
  EXPECT_EQ(FromDomain(kPortalDomainStaging).value(), kStaging);
  EXPECT_EQ(FromDomain(kAccountsDomainStaging).value(), kStaging);
  EXPECT_EQ(FromDomain(kAssetsDomainStaging).value(), kStaging);
}

TEST(EnvironmentsTest, FromDomainDev) {
  EXPECT_EQ(FromDomain(kPortalDomainDev).value(), kDev);
  EXPECT_EQ(FromDomain(kAccountsDomainDev).value(), kDev);
  EXPECT_EQ(FromDomain(kAssetsDomainDev).value(), kDev);
}

TEST(EnvironmentsTest, FromDomainInvalid) {
  absl::StatusOr<std::string> result = FromDomain("invalid_domain");
  EXPECT_FALSE(result.ok());
  EXPECT_EQ(result.status().code(), absl::StatusCode::kInvalidArgument);
  EXPECT_THAT(result.status().message(),
              HasSubstr("unknown domain: invalid_domain"));
}

TEST(EnvironmentsTest, FromProjectProd) {
  EXPECT_EQ(FromProject(kPortalProjectProd).value(), kProd);
  EXPECT_EQ(FromProject(kAccountsProjectProd).value(), kProd);
  EXPECT_EQ(FromProject(kAssetsProjectProd).value(), kProd);
}

TEST(EnvironmentsTest, FromProjectStaging) {
  EXPECT_EQ(FromProject(kPortalProjectStaging).value(), kStaging);
  EXPECT_EQ(FromProject(kAccountsProjectStaging).value(), kStaging);
  EXPECT_EQ(FromProject(kAssetsProjectStaging).value(), kStaging);
}

TEST(EnvironmentsTest, FromProjectDev) {
  EXPECT_EQ(FromProject(kPortalProjectDev).value(), kDev);
  EXPECT_EQ(FromProject(kAccountsProjectDev).value(), kDev);
  EXPECT_EQ(FromProject(kAssetsProjectDev).value(), kDev);
}

TEST(EnvironmentsTest, FromProjectInvalid) {
  absl::StatusOr<std::string> result = FromProject("invalid_project");
  EXPECT_FALSE(result.ok());
  EXPECT_EQ(result.status().code(), absl::StatusCode::kInvalidArgument);
  EXPECT_THAT(result.status().message(),
              HasSubstr("unknown project: invalid_project"));
}

TEST(EnvironmentsTest, FromComputeProjectProd) {
  EXPECT_EQ(FromComputeProject("some_other_project"), kProd);
}

TEST(EnvironmentsTest, PortalDomain) {
  EXPECT_EQ(PortalDomain(kProd), kPortalDomainProd);
  EXPECT_EQ(PortalDomain(kStaging), kPortalDomainStaging);
  EXPECT_EQ(PortalDomain(kDev), kPortalDomainDev);
  EXPECT_EQ(PortalDomain("invalid_env"), "");
}

TEST(EnvironmentsTest, AccountsDomain) {
  EXPECT_EQ(AccountsDomain(kProd), kAccountsDomainProd);
  EXPECT_EQ(AccountsDomain(kStaging), kAccountsDomainStaging);
  EXPECT_EQ(AccountsDomain(kDev), kAccountsDomainDev);
  EXPECT_EQ(AccountsDomain("invalid_env"), "");
}

TEST(EnvironmentsTest, AccountsProjectFromEnv) {
  EXPECT_EQ(AccountsProjectFromEnv(kProd), kAccountsProjectProd);
  EXPECT_EQ(AccountsProjectFromEnv(kStaging), kAccountsProjectStaging);
  EXPECT_EQ(AccountsProjectFromEnv(kDev), kAccountsProjectDev);
  EXPECT_EQ(AccountsProjectFromEnv("invalid_env"), "");
}

TEST(EnvironmentsTest, AccountsProjectFromProject) {
  EXPECT_EQ(AccountsProjectFromProject(kPortalProjectProd),
            kAccountsProjectProd);
  EXPECT_EQ(AccountsProjectFromProject(kPortalProjectStaging),
            kAccountsProjectStaging);
  EXPECT_EQ(AccountsProjectFromProject(kPortalProjectDev), kAccountsProjectDev);

  EXPECT_EQ(AccountsProjectFromProject(kAccountsProjectProd),
            kAccountsProjectProd);
  EXPECT_EQ(AccountsProjectFromProject(kAccountsProjectStaging),
            kAccountsProjectStaging);
  EXPECT_EQ(AccountsProjectFromProject(kAccountsProjectDev),
            kAccountsProjectDev);

  EXPECT_EQ(AccountsProjectFromProject(kAssetsProjectProd),
            kAccountsProjectProd);
  EXPECT_EQ(AccountsProjectFromProject(kAssetsProjectStaging),
            kAccountsProjectStaging);
  EXPECT_EQ(AccountsProjectFromProject(kAssetsProjectDev), kAccountsProjectDev);

  EXPECT_EQ(AccountsProjectFromProject("some_other_project"),
            kAccountsProjectProd);
}

TEST(EnvironmentsTest, AssetsDomain) {
  EXPECT_EQ(AssetsDomain(kProd), kAssetsDomainProd);
  EXPECT_EQ(AssetsDomain(kStaging), kAssetsDomainStaging);
  EXPECT_EQ(AssetsDomain(kDev), kAssetsDomainDev);
  EXPECT_EQ(AssetsDomain("invalid_env"), "");
}

TEST(EnvironmentsTest, AssetsProject) {
  EXPECT_EQ(AssetsProject(kProd), kAssetsProjectProd);
  EXPECT_EQ(AssetsProject(kStaging), kAssetsProjectStaging);
  EXPECT_EQ(AssetsProject(kDev), kAssetsProjectDev);
  EXPECT_EQ(AssetsProject("invalid_env"), "");
}

TEST(EnvironmentsTest, All) {
  EXPECT_THAT(kAll, testing::UnorderedElementsAre(kProd, kStaging, kDev));
}

}  // namespace environments

File: intrinsic/config/environments_test.py
# Copyright 2023 Intrinsic Innovation LLC

from absl.testing import absltest
from intrinsic.config import environments


class EnvironmentsTest(absltest.TestCase):

  def test_all_environments_constant(self):
    self.assertEqual(environments.ALL, ["prod", "staging", "dev"])
    self.assertIn(environments.PROD, environments.ALL)
    self.assertIn(environments.STAGING, environments.ALL)
    self.assertIn(environments.DEV, environments.ALL)

  def test_from_domain_valid(self):
    self.assertEqual(
        environments.from_domain(environments.ACCOUNTS_DOMAIN_PROD),
        environments.PROD,
    )
    self.assertEqual(
        environments.from_domain(environments.PORTAL_DOMAIN_STAGING),
        environments.STAGING,
    )
    self.assertEqual(
        environments.from_domain(environments.ASSETS_DOMAIN_DEV),
        environments.DEV,
    )

  def test_from_domain_invalid(self):
    with self.assertRaisesRegex(
        ValueError, "Unknown domain: invalid.domain.com"
    ):
      environments.from_domain("invalid.domain.com")

  def test_from_project_valid(self):
    self.assertEqual(
        environments.from_project(environments.ACCOUNTS_PROJECT_PROD),
        environments.PROD,
    )
    self.assertEqual(
        environments.from_project(environments.PORTAL_PROJECT_STAGING),
        environments.STAGING,
    )
    self.assertEqual(
        environments.from_project(environments.ASSETS_PROJECT_DEV),
        environments.DEV,
    )

  def test_from_project_invalid(self):
    with self.assertRaisesRegex(ValueError, "Unknown project: invalid-project"):
      environments.from_project("invalid-project")

  def test_from_compute_project(self):
    # from_compute_project currently delegates to from_project
    self.assertEqual(
        environments.from_compute_project(environments.PORTAL_PROJECT_PROD),
        environments.PROD,
    )
    self.assertEqual(
        environments.from_compute_project("invalid-compute-project"),
        environments.PROD,
    )

  def test_portal_domain(self):
    self.assertEqual(
        environments.portal_domain(environments.PROD),
        environments.PORTAL_DOMAIN_PROD,
    )
    self.assertEqual(
        environments.portal_domain(environments.STAGING),
        environments.PORTAL_DOMAIN_STAGING,
    )
    self.assertEqual(
        environments.portal_domain(environments.DEV),
        environments.PORTAL_DOMAIN_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.portal_domain("invalid_env")

  def test_portal_project(self):
    self.assertEqual(
        environments.portal_project(environments.PROD),
        environments.PORTAL_PROJECT_PROD,
    )
    self.assertEqual(
        environments.portal_project(environments.STAGING),
        environments.PORTAL_PROJECT_STAGING,
    )
    self.assertEqual(
        environments.portal_project(environments.DEV),
        environments.PORTAL_PROJECT_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.portal_project("invalid_env")

  def test_accounts_domain(self):
    self.assertEqual(
        environments.accounts_domain(environments.PROD),
        environments.ACCOUNTS_DOMAIN_PROD,
    )
    self.assertEqual(
        environments.accounts_domain(environments.STAGING),
        environments.ACCOUNTS_DOMAIN_STAGING,
    )
    self.assertEqual(
        environments.accounts_domain(environments.DEV),
        environments.ACCOUNTS_DOMAIN_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.accounts_domain("invalid_env")

  def test_accounts_project_from_env(self):
    self.assertEqual(
        environments.accounts_project_from_env(environments.PROD),
        environments.ACCOUNTS_PROJECT_PROD,
    )
    self.assertEqual(
        environments.accounts_project_from_env(environments.STAGING),
        environments.ACCOUNTS_PROJECT_STAGING,
    )
    self.assertEqual(
        environments.accounts_project_from_env(environments.DEV),
        environments.ACCOUNTS_PROJECT_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.accounts_project_from_env("invalid_env")

  def test_accounts_project_from_project(self):
    self.assertEqual(
        environments.accounts_project_from_project(
            environments.PORTAL_PROJECT_PROD
        ),
        environments.ACCOUNTS_PROJECT_PROD,
    )
    self.assertEqual(
        environments.accounts_project_from_project(
            environments.ASSETS_PROJECT_STAGING
        ),
        environments.ACCOUNTS_PROJECT_STAGING,
    )
    self.assertEqual(
        environments.accounts_project_from_project("invalid-project"),
        environments.ACCOUNTS_PROJECT_PROD,
    )

  def test_assets_domain(self):
    self.assertEqual(
        environments.assets_domain(environments.PROD),
        environments.ASSETS_DOMAIN_PROD,
    )
    self.assertEqual(
        environments.assets_domain(environments.STAGING),
        environments.ASSETS_DOMAIN_STAGING,
    )
    self.assertEqual(
        environments.assets_domain(environments.DEV),
        environments.ASSETS_DOMAIN_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.assets_domain("invalid_env")

  def test_assets_project(self):
    self.assertEqual(
        environments.assets_project(environments.PROD),
        environments.ASSETS_PROJECT_PROD,
    )
    self.assertEqual(
        environments.assets_project(environments.STAGING),
        environments.ASSETS_PROJECT_STAGING,
    )
    self.assertEqual(
        environments.assets_project(environments.DEV),
        environments.ASSETS_PROJECT_DEV,
    )
    with self.assertRaisesRegex(ValueError, "Unknown environment: invalid_env"):
      environments.assets_project("invalid_env")


if __name__ == "__main__":
  absltest.main()
